src/key_schedule.rs

```diff
@@ -48,104 +36,85 @@ impl SecretKind {
             current_server_traffic_secret: Vec::new(),
             current_client_traffic_secret: Vec::new(),
             current_exporter_secret: Vec::new(),
         }
     }
 
+    #[inline]
+    pub fn algorithm(&self) -> hkdf::Algorithm { self.algorithm }
+
+    pub fn new_with_empty_secret(algorithm: hkdf::Algorithm) -> KeySchedule {
+        let zeroes = [0u8; digest::MAX_OUTPUT_LEN];
+        let hash_len = algorithm.hmac_algorithm().digest_algorithm().output_len;
+        Self::new(algorithm, &zeroes[..hash_len])
+    }
+
     /// Input the empty secret.
     pub fn input_empty(&mut self) {
         let zeroes = [0u8; digest::MAX_OUTPUT_LEN];
-        let hash_len = self.hash.output_len;
+        let hash_len = self.algorithm.hmac_algorithm().digest_algorithm().output_len;
         self.input_secret(&zeroes[..hash_len]);
     }
 
     /// Input the given secret.
     pub fn input_secret(&mut self, secret: &[u8]) {
-        if self.need_derive_for_extract {
-            let derived = self.derive_for_empty_hash(SecretKind::DerivedSecret);
-            self.current = hmac::Key::new(self.hmac_alg,
-                                          &derived);
-        }
-        self.need_derive_for_extract = true;
-        let new = _hkdf_extract(&self.current, secret);
-        self.current = new
+        let salt: hkdf::Salt =
+            self.derive_for_empty_hash(self.algorithm, SecretKind::DerivedSecret);
+        self.current = salt.extract(secret);
     }
 
     /// Derive a secret of given `kind`, using current handshake hash `hs_hash`.
-    pub fn derive(&self, kind: SecretKind, hs_hash: &[u8]) -> Vec<u8> {
-        debug_assert_eq!(hs_hash.len(), self.hash.output_len);
+    pub fn derive<T, L>(&self, key_type: L, kind: SecretKind, hs_hash: &[u8]) -> T
+        where
+            T: for <'a> From<hkdf::Okm<'a, L>>,
+            L: hkdf::KeyType,
+    {
+        hkdf_expand(&self.current, key_type, kind.to_bytes(), hs_hash)
+    }
 
-        _hkdf_expand_label_vec(&self.current,
-                               kind.to_bytes(),
-                               hs_hash,
-                               self.hash.output_len)
+    pub fn derive_bytes(&self, kind: SecretKind, hs_hash: &[u8]) -> Vec<u8> {
+        let payload: PayloadU8 = hkdf_expand(
+            &self.current,
+            PayloadU8Len(self.algorithm.len()),
+            kind.to_bytes(),
+            hs_hash);
+        payload.into_inner()
     }
 
+
     /// Derive a secret of given `kind` using the hash of the empty string
     /// for the handshake hash.  Useful only for
     /// `SecretKind::ResumptionPSKBinderKey` and
     /// `SecretKind::DerivedSecret`.
-    pub fn derive_for_empty_hash(&self, kind: SecretKind) -> Vec<u8> {
-        let mut empty_hash = [0u8; digest::MAX_OUTPUT_LEN];
-        empty_hash[..self.hash.output_len]
-            .clone_from_slice(digest::digest(self.hash, &[]).as_ref());
-
-        self.derive(kind, &empty_hash[..self.hash.output_len])
```

Comment:
This code looked weird to me. Maybe I'm misunderstanding it. Please double-check.

Reply:
Yes, it was weird. For no good reason.